File: custom built data structures/customHashTable.py
class HashTable():
    def __init__(self, size):
        self.data = [None] * size

    def hash(self, key):
        return hash(key) % len(self.data)  # Python's built-in hash function
    
    def set(self, key, value):
        address = self.hash(key)
        item = [key, value]

        if self.data[address] is None:
            self.data[address] = []

        self.data[address].append(item)

        return self.data
    
    def get(self, key):
        address = self.hash(key)
        currentBucket = self.data[address]

        if currentBucket:
            for i in range(len(currentBucket)):
                if currentBucket[i][0] == key:
                    return currentBucket[i][1]

        return None

newHashTable = HashTable(50)


newHashTable.set('grapes', 10000)
newHashTable.set('apples', 1000)
newHashTable.set('mangoes', 1000)
newHashTable.set('beer', 100)

print(newHashTable)

print(newHashTable.get('grapes'))
print(newHashTable.get('apples'))
print(newHashTable.get('mangoes'))
print(newHashTable.get('beer'))

File: custom-built-data-structures/javascript/customHashTable.js
class HashTable {
  constructor(size) {
    this.data = new Array(size);
  }

  _hash(key) {
    let hash = 0;
    for (let i = 0; i < key.length; i++) {
      hash = (hash + key.charCodeAt(i) * i) % this.data.length;
    }
    return hash;
  }

  set(key, value) {
    let address = this._hash(key);
    let item = [key, value];
    if (!this.data[address]) {
      this.data[address] = [];
    }
    this.data[address].push(item);

    return this.data;
  }

  get(key) {
    let address = this._hash(key);
    let currentBucket = this.data[address];
    if (currentBucket) {
      for (let i = 0; i < currentBucket.length; i++) {
        if (currentBucket[i][0] == key) {
          return currentBucket[i][1];
        }
      }
    }
    return undefined;
  }
}

const myHashTable = new HashTable(50);
myHashTable.set("grapes", 10000);
// myHashTable.get("grapes");
myHashTable.set("apples", 9);
myHashTable.get("apples");

console.log(myHashTable);
// expected output

// HashTable { data: [ <50 empty items> ["grapes", 10000],["apples", 5000],["mangoes", 55] ] }
